torch/distributed/_tensor/dispatch.py

```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
     suggested_input_schema = output_sharding.schema_suggestions[0]
     needs_redistribute = suggested_input_schema is not op_schema
 
+    def _arg_type_is_primitive(arg_type):
+        _python_primitive_types = [
+            "bool",
+            "int",
+            "float",
+            "double",
+        ]
+        return arg_type in _python_primitive_types
+
     if mesh is not None and mesh.get_coordinate() is None:
-        # if we are on a non-participating device, we simply return
-        # an empty tensor for now.
-        # TODO: what if the op returns a non-tensor value, what if
-        # the op returns a list of tensors, we need to figure out
-        # a consistent way to handle that, and also need to figure
-        # out if we should communicate the result to non-participating
-        # ranks (i.e. a.sum() -> scalar, maybe we should set to 0)
-        local_results = torch.tensor([])
+        # For a non-participating device, we do:
+        # 1. if the return type is scalar, all gather the local result
+        # from participating devices, and reduce on the list of results.
+        # For bool type, we by default use AND to reduce;
+        # For numeric type, we by default use ADD to reduce;
+        # We can extend this logic for specific ops if necessary.
+        # 2. if the return type is Tensor, return empty tensor with
+        # correct dtype.
+        # 3. if the return type is List[Tensor], return a list of empty
+        # tensors with correct dtype.
+
+        def _default_tensor(spec: Optional[DTensorSpec]):
+            assert spec is not None, (
+                f"expect DTensorSpec object but got {spec}"
+            )
+            if spec.tensor_meta is not None:
+                shape = spec.tensor_meta.shape
+                dtype = spec.tensor_meta.dtype
+                if len(shape) == 0:
+                    # scalar tensor
+                    return torch.zeros(
+                        (),
+                        dtype=dtype,
+                    )
+                else:
+                    # non-scalar tensor
+                    return torch.tensor(
+                        [],
```

Comment:
the indention here seems.. weird, it should definitely fit in a single line, you might want to check if your VSCode have some issue (maybe you have some ufmt that format files to line-length 80 only